tune-threads/tune-threads.lisp

```diff
@@ -0,0 +1,169 @@
+;;;; tune-threads/tune-threads.lisp
+;;;;
+;;;; Author: John Lapeyre
+
+(in-package #:qvm-tune-threads)
+
+(defstruct pre-timing
+  "Data characterizing coarse timing of quil program."
+  (init-num-trials 1 :read-only t)
+  (num-trials 1 :read-only t)
+  (trial-time 1.0 :read-only t))
+
+(defparameter *minimum-target-time* 0.1
+  "The minimum required execution time for a block of repeated trials of a quil program.
+The number of trials will be increased until this target time is exceeded.")
+
+(defun time-quil-prog-raw (q &key (num-trials 1) (num-threads 4))
+  "Return the execution time in seconds of the quil program in qvm Q."
+  (qvm::prepare-for-parallelization num-threads)
+  (simple-time num-trials (run-and-reset-qvm q)))
+
+(defun run-pre-timing (q num-threads init-num-trials)
+  "Perform coarse timing of program in qvm Q."
+  (loop :for num-pre-trials := init-num-trials :then (* num-pre-trials 2)
+        :for trial-time := (time-quil-prog-raw q :num-threads num-threads :num-trials num-pre-trials)
+        :while (< trial-time *minimum-target-time*) ; avoid divide by zero and times very close to zero
+        :finally (return (make-pre-timing :init-num-trials init-num-trials :num-trials num-pre-trials :trial-time trial-time))))
+
+(defun num-trials-in-target-time (pt target-time)
+  "Use coarse timing data in PT to determine the number of trials required to consume TARGET-TIME seconds.
+The latter time is used for more precise timing."
+  (1+ (floor target-time (/ (pre-timing-trial-time pt) (pre-timing-num-trials pt)))))
+
+(defstruct timing
+  "Data characterizing precise (relative to coarse timing) timing of a quil program.
+All slots are meant to be read only, and have no default values."
+  pre-timing
+  num-trials
+  num-threads
```

Comment:
Yes, I thought about it, but did not add it because it was not needed since `num-qubits` is a hash key. But, it might be a good idea to add the slot.